### DISCUSSION TITLE
Issue with multiple sink levels in one logger

### DISCUSSION THREAD

**Discussion by ghost** [gabime/spdlog#2107] [Q&A] [answered]
I'm having issues using multiple sinks, with different levels, in one logger (using the default logger in general as well as the logger itself).  

Using the examples from the wiki -  I put the code below together...  When testing the logger with the included sinks, neither message prints out.  To either the console or the file.  From what I've read, I am under the impression that this should at least print out in the console, right? 

I recently added the preprocessor definition at the top to try to set a baseline level... it's not doing much.

This is being called by another file in the same project, would that cause any issues?

```
#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE
#include "spdlog/spdlog.h"
#include "spdlog/sinks/basic_file_sink.h" // File logger
#include "spdlog/sinks/stdout_color_sinks.h" // Multi-colored console logger

...

void logging::initialize_spdlog() {
    std::vector<spdlog::sink_ptr> sinks;

    // File logger
    auto file_sink = std::make_shared<spdlog::sinks::basic_file_sink_mt>(
            utilities::get_log_file_path()
        );
    file_sink->set_pattern("%Y-%m-%d %T,%e : %-40!! : %-7l : %v");
    //%-20!s :
    sinks.push_back(file_sink);
    file_sink->set_level(spdlog::level::info);

    // Console logger
    auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
    console_sink->set_pattern("%T,%e : %-7l : %v");
    console_sink->set_level(spdlog::level::trace);
    sinks.push_back(console_sink);

    auto daq_logger = std::make_shared<spdlog::logger>("logger", begin(sinks), end(sinks));
    // don't think this is needed, but nothing will work anyway
    spdlog::register_logger(daq_logger);

    spdlog::set_default_logger(daq_logger);
    // test
    daq_logger->debug("logger");
    spdlog::debug("default logger");
}
```

**Comment by tt4g** [comment, marked as answer]
`spdlog::logger` also has log level (by default info).

https://github.com/gabime/spdlog/blob/v1.9.2/include/spdlog/logger.h#L314

No log messages will be passed to sink if they are filtered at the logger log level.

https://github.com/gabime/spdlog/blob/v1.9.2/include/spdlog/logger.h#L127-L132

You have to set log level of `daq_logger` to `spdlog::level::trace`.

```diff
     auto daq_logger = std::make_shared<spdlog::logger>("logger", begin(sinks), end(sinks));
+  daq_logger ->set_level(spdlog::level::trace);
     // don't think this is needed, but nothing will work anyway
     spdlog::register_logger(daq_logger);
```

**Reply by ghost** [reply]
Ohhhh, ok. That makes sense!

Also, if I use a spdlog macro will it use the sink levels that are set? I wanted to have similar functionality with file info when logged.

**Reply by tt4g** [reply]
The logging macro such as `SPDLOG_ACTIVE_LEVEL`, `SPDLOG_LOGGER_TRACE()` and `SPDLOG_TRACE` is a different function from the log level.
The logging macro is used to remove the code less than or equal to the level specified by the `SPDLOG_ACTIVE_LEVEL` macro when pre-processing.
See Wiki: https://github.com/gabime/spdlog/wiki/0.-FAQ#how-to-remove-all-debug-statements-at-compile-time-

**Reply by ghost** [reply]
That's what I thought. Thanks!

**Reply by ghost** [reply]
I found that if I set the `SPDLOG_ACTIVE_LEVEL` to `trace` and set the logger above as the default, I get the intended functionality.
